### PR TITLE
feat: multi collateral warp routes

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

Adds multicollateral warp route support to the CLI. Dedupes SDK warp route types and zod schemas (via type inference).

### Drive-by changes

- Removes scattered ICA logic from deployers in favor of config time resolution

### Related issues

- Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3639

### Backward compatibility

- No

### Testing

- E2E Tests

### CODE REVIEW

File: typescript/sdk/src/token/schemas.ts [hunk 1]
```diff
@@ -63,12 +48,36 @@ export const TokenConfigSchema = z.discriminatedUnion('type', [
...
-  TokenRouterConfigSchema,
-);
+export const WarpRouteDeployConfigSchema = z
+  .record(TokenRouterConfigSchema)
+  .refine((configMap) => {
+    const entries = Object.entries(configMap);
+    return (
+      entries.some(
+        ([_, config]) =>
+          CollateralConfigSchema.safeParse(config).success ||
```
Comment (ltyu) [inline typescript/sdk/src/token/schemas.ts:62]: can we use `isCompliant`, or one of the helpers that you created below?

File: typescript/sdk/src/token/schemas.ts [hunk 2]
```diff
@@ -63,12 +48,36 @@ export const TokenConfigSchema = z.discriminatedUnion('type', [
...
-export const WarpRouteDeployConfigSchema = z.record(
-  z.string(),
-  TokenRouterConfigSchema,
-);
+export const WarpRouteDeployConfigSchema = z
+  .record(TokenRouterConfigSchema)
+  .refine((configMap) => {
+    const entries = Object.entries(configMap);
+    return (
+      entries.some(
```
Comment (ltyu) [inline typescript/sdk/src/token/schemas.ts:60]: would this work if the configMap had a mix of valid and invalid configs?

**Review by ltyu — COMMENTED** [review]
Looks pretty good upon first review. just some quick questions

File: typescript/sdk/src/token/schemas.ts (hunk 2 above)
Comment (ltyu) [reply typescript/sdk/src/token/schemas.ts:60]: quick failing test for you. we're probably not expecting this combo in prod?

```typescript
it.only('should throw if one of the configs is invalid', async () => {
    const collateral = {
      arbitrum: {
        ...config.arbitrum,
        type: TokenType.collateral,
        token: SOME_ADDRESS,
      },
      op: {
        ...config.arbitrum,
        type: TokenType.synthetic,
        // symbol: 'symbol',
      },
    };
    expect(WarpRouteDeployConfigSchema.safeParse(collateral).success).to.be
      .false;
});

```

File: typescript/cli/src/deploy/warp.ts [hunk 3]
```diff
@@ -91,111 +83,13 @@ export async function runWarpRouteDeploy({
...
-  log('Assembling token configs');
-  const chainAddresses = await registry.getAddresses();
-  const owner = await signer.getAddress();
-  const requiredRouterFields: Array<keyof ConnectionClientConfig> = ['mailbox'];
-  const remotes: string[] = [];
-
-  /// @dev This will keep track of the base collateral metadata which can get overwritten if there are multiple collaterals.
-  /// These 'base' variables are used to derive synthetic fields
-  /// @todo Remove this artifact when multi-collateral is enabled
-  let baseChainName = '';
```
Comment (ltyu) [inline typescript/cli/src/deploy/warp.ts:111]: dope

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts [hunk 4]
```diff
@@ -44,34 +44,32 @@ export class EvmERC20WarpRouteReader {
...
+        token = await HypERC20Collateral__factory.connect(
+          address,
+          this.provider,
+        ).wrappedToken();
+        break;
+      case TokenType.synthetic:
+        token = address;
+        break;
+      default:
+        throw new Error(`Invalid token type: ${type}`);
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:63]: I believe @ltyu 's work will land improvements here soon

File: typescript/sdk/src/token/schemas.ts [hunk 5]
```diff
@@ -63,12 +48,36 @@ export const TokenConfigSchema = z.discriminatedUnion('type', [
...
+        ([_, config]) =>
+          CollateralConfigSchema.safeParse(config).success ||
+          NativeConfigSchema.safeParse(config).success,
+      ) ||
+      entries.filter(
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply typescript/sdk/src/token/schemas.ts:65]: nice, didnt know about this

File: typescript/sdk/src/token/schemas.ts (hunk 1 above)
Comment (yorhodes) [reply typescript/sdk/src/token/schemas.ts:62]: yes :)

File: typescript/cli/src/deploy/warp.ts [hunk 6]
```diff
@@ -210,80 +104,63 @@ async function executeDeploy(params: DeployParams) {
...
-      `Unsupported token: ${config.type}. Consider setting token metadata in your deployment config.`,
-    );
-  }
-}
+const DEFAULT_METADATA = {
+  name: 'unknown',
+  symbol: 'unknown',
+  totalSupply: '0',
+  decimals: 18,
+};
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply typescript/cli/src/deploy/warp.ts:135]: sgtm

File: typescript/sdk/src/token/schemas.ts (hunk 2 above)
Comment (yorhodes) [reply typescript/sdk/src/token/schemas.ts:60]: this is a valid config
the token metadata for the synthetic will be derived from the collateral token

File: typescript/sdk/src/token/schemas.ts [hunk 7]
```diff
@@ -63,12 +48,36 @@ export const TokenConfigSchema = z.discriminatedUnion('type', [
...
+        ([_, config]) => TokenMetadataSchema.safeParse(config).success,
+      ).length === entries.length
+    );
+  }, `Config must include Native or Collateral OR all synthetics must define token metadata`);
+
+export type TokenRouterConfig = z.infer<typeof TokenRouterConfigSchema>;
+export type NativeConfig = z.infer<typeof NativeConfigSchema>;
+export type CollateralConfig = z.infer<typeof CollateralConfigSchema>;
+
+function isCompliant<S extends Zod.Schema>(schema: S) {
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply typescript/sdk/src/token/schemas.ts:75]: I view this util as useful for type narrowing, as opposed to getting the boolean result

File: typescript/cli/src/config/warp.ts [hunk 8]
```diff
@@ -40,75 +60,68 @@ export async function createWarpRouteDeployConfig({
...
-    result[chain] = {
-      type: TokenType.synthetic,
-    };
-  });
-
   if (isValidWarpRouteDeployConfig(result)) {
     logGreen(`Warp Route config is valid, writing to file ${outPath}`);
     writeYamlOrJson(outPath, result);
   } else {
     errorRed(
```
Comment (ltyu) [inline typescript/cli/src/config/warp.ts:121]: is this error useful? isn't this function going to create the yaml?

if the function fails to create the yaml, how would i compare it to the example?

if we remove this error, then this conditional can be removed, and we can just `.parse`, and then `writeYamlOrJson`

File: typescript/cli/src/deploy/warp.ts [hunk 9]
```diff
@@ -63,26 +59,22 @@ export async function runWarpRouteDeploy({
...
   };
 
   logBlue('Warp route deployment plan');
 
   await runDeployPlanStep(deploymentParams);
-  await runPreflightChecks({
-    ...deploymentParams,
+  await runPreflightChecksForChains({
+    context,
+    chains: Object.keys(warpRouteConfig),
```
Comment (ltyu) [inline typescript/cli/src/deploy/warp.ts:72]: nit: can move the `const chains = ...` line below higher and use it instead :)

**Review by jmrossy — APPROVED** [review]

File: typescript/sdk/src/token/types.ts [hunk 10]
```diff
@@ -1,9 +1,11 @@
 import { z } from 'zod';
 
 import {
+  TokenMetadataSchema,
   TokenRouterConfigSchema,
   WarpRouteDeployConfigSchema,
 } from './schemas.js';
 
+export type TokenMetadata = z.infer<typeof TokenMetadataSchema>;
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply typescript/sdk/src/token/types.ts:9]: lets do this wholesale

File: typescript/sdk/src/token/deploy.hardhat-test.ts [hunk 11]
```diff
@@ -2,38 +2,31 @@ import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers.js';
...
-import { ChainMap } from '../types.js';
 
-import { EvmERC20WarpRouteReader } from './EvmERC20WarpRouteReader.js';
-import {
-  HypERC20CollateralConfig,
-  HypERC20Config,
-  TokenConfig,
-  TokenType,
-} from './config.js';
+import { EvmWarpRouteReader } from './EvmERC20WarpRouteReader.js';
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply typescript/sdk/src/token/deploy.hardhat-test.ts:15]: thats why I renamed it :)
it works with 721 too

**Pull request merged**